src/GraphQL/Types/Scalars/IdGraphType.cs

```diff
@@ -23,31 +23,39 @@ public IdGraphType()
             //    "as `\"4\"`) or integer (such as `4`) input value will be accepted as an `ID`.";
         }
 
-        /// <inheritdoc/>
-        public override object Serialize(object value) => value.ToString();
-
         /// <inheritdoc/>
         public override object ParseLiteral(IValue value) => value switch
         {
             StringValue str => ParseValue(str.Value),
             IntValue num => num.Value,
             LongValue longVal => longVal.Value,
+            BigIntValue bigIntValue => bigIntValue.Value,
             _ => null,
         };
 
-        /// <inheritdoc/>
-        public override object ParseValue(object value) => value.ToString().Trim(' ', '"');
-
         /// <inheritdoc/>
         public override bool CanParseLiteral(IValue value) => value switch
         {
             StringValue _ => true,
             IntValue _ => true,
             LongValue _ => true,
+            BigIntValue _ => true,
+            NullValue _ => true,
             _ => false
         };
 
         /// <inheritdoc/>
-        public override IValue ToAST(object value) => new StringValue(value.ToString());
+        public override object ParseValue(object value) => value switch
+        {
+            int _ => value,
+            long _ => value,
+            _ => value?.ToString().Trim(' ', '"')
+        };
```

Comment:
So.... when is `int` / `long` preserved, and when is it converted to strings?  Always convert to string and trim?  Why trim?  Why convert to string?  Not sure here.

Reply:
> The ID scalar graph type represents a **string** identifier, not intended to be human-readable. 
\+
> When expected as an **input** type, any string or integer input value will be accepted as an ID.

See http://spec.graphql.org/draft/#sec-ID

So ...
> when is it converted to strings?

When serializing into response.

> Why convert to string?

Because spec requires it.

>  Why trim?

It seems to me that this is some kind of design flaw. Spec says nothing about trimming. Try run tests without trim.

Reply:
Right; I understand that it should serialize to strings when serializing values.  I mean, when being parsed, why, or why not, convert to strings?  Perhaps we should always convert literals and values to strings, or perhaps never (and let GetArgument handle this).  Or perhaps we never convert values to strings, allowing any type (float, decimal, sbyte, whatever).

I'll try removing Trim.

Reply:
breaks these tests:
```cs
        [Fact]
        public void parse_value_quoted_string_to_identifier()
        {
            _type.ParseValue("\"12345\"").ShouldBe("12345");
        }

        [Fact]
        public void parse_literal_string_value_quoted_string_to_identifier()
        {
            _type.ParseLiteral(new StringValue("\"12345\"")).ShouldBe("12345");
        }
```
(which still don't seem to make sense to me)

Reply:
I'm going to remove these two tests as well.  Lmk if not

Reply:
>  I mean, when being parsed, why, or why not, convert to strings? Perhaps we should always convert literals and values to strings, or perhaps never (and let GetArgument handle this). 

I think that `ID` as a input type _can_ be used in different ways depending on the type of the variable/literal. Type of input   is an additional source of information besides the actual value. Imagine two options: `"123"` and `123`

```graphql
query {
  user(id: ID!)
}
```

And imagine this `User` CLR type:
```csharp
public class User
{
   public int InternalID { get; set; }
   public string ExternalID { get; set; }
}
```

Resolver:
```csharp
context => {
  if (context.Arguments["id"] is int id)
  {
     // search by internal id
  }
  else if (context.Arguments["id"] is string s)
  {
    // search by external id
  }
}
```

> I'm going to remove these two tests as well. Lmk if not

OK. Even if we did not understand the purpose of `Trim`, then after the release of v4 we will surely receive some feedback. If necessary, we can return the old behavior. Also ping @joemcbride for advice.

Reply:
We can always add `Trim` back in without a breaking change.  Similar with other restrictions we have made for scalars.  Also I think we should most definitely provide a sample of overriding a built-in scalar within the migration documentation.